### DISCUSSION TITLE
Support for array of objects

### DISCUSSION THREAD

**Discussion by frankist** [CLIUtils/CLI11#887] [Q&A]
Hi,

I have a quite complicated configuration setup, consisting of a relatively large config file, and I like to override the parameters in this config file with parameters I pass through the command line. CLI11 seems perfect for this.

However, I run into a problem, which I don't know how to solve. My configuration file contains some arrays of objects (lists where each item has its own subfields) and I don't know how to parse these arrays via CLI11. Is this feature supported at all?

I thought of converting my array of objects into several subcommands, but that would require specifying a limit for the array, and it would make the `--help` look ugly. e.g.
```
Subcommands:
  item0 ...
  item1 ...
  ...
```

If this feature is not supported, is there a plan to do so? I wouldn't even mind if these arrays of objects were not configurable through the command line, as long as they would get parsed by CLI11 when set in the configuration file.

**Comment by phlptp** [comment]
I think I would need to see an example of the config file you want to support to answer for certain.  Sub commands or option groups can be triggered multiple times to handle an array like structure.  Or it might make sense to make your own parser customization to handle things a little cleaner.

**Reply by frankist** [reply]
Thanks for the reply.

For improved clarification, let's say I have:
```c++
struct item {
  int         field1;
  std::string field2;
};
```
And I have an application configuration struct like this:
```
struct config {
  std::vector<item> items;
};
```
I can have an arbitrary number of `items`, so my config (in json) could look like this:
```
{
  "items": [ 
    { "field1": 1, "field2": "hello" },
    { "field1": 2, "field2": "world" }
  ]
}
```
How would I write the CLI11 subcommands and options for this configuration structure? One approach I thought about was doing something like this:
```c++
app.add_option_function<std::vector<std::string>>("--items", [&items_vec](const std::vector<std::string>& text_items) {
  items_vec.resize(text_items.size());
  for(unsigned i = 0; i != text_items.size(); ++i) {
    CLI::App subapp("items");
    subapp.config_formatter(...);
    // define the schema for subapp
    std::istringstream ss(text_items[i]);
    subapp.parse_from_stream(ss);
  }
}, "description");
```
Basically I create a new CLI11 subapp to parse each item of the array. It works, but it is not very elegant and leads to a very weird `--help` output that looks like this:
```
Options:
  --items TEXT ...            description
```
Maybe I am missing an obviously simpler and cleaner way to do this?

**Reply by senr80** [reply]
Hi,

I have a very similar use case: a config file with a recurring set of parameters to configure an arbitrary number of sessions (although without the need to override them on the command line). The format is (kind of) TOML and for this sessions it has entries like this (for historical reasons):

```
foo.0.someField = bar
foo.0.anotherField = 123
...
```
I would like to use CLI11 to parse this into a vector of structs as this also gives a nice documentation in the code about the parameters required to setup these sessions (in contrary to the way it is solved atm).

So far, my solution is that I customized `ConfigTOML::from_config()` to create `ConfigItem`s for such parameters in a way that `foo` becomes the value of the item's `name` and the everything else is moved to the `input` attribute like `['0', 'someField', '=', 'bar']`. 

Then, I derived a class from `App` to get something like an option group. The constructor takes a name, a reference to the target variable, a description and an `App` pointer to a parent. It adds an option to the parent using the name and the description and a lambda function unpacks the `ConfigItem::input` that arrived through the option's `results_` and writes the right element to the proper index of the target variable. The lambda also calls `clear()` on the option, otherwise subsequent items aren't considered anymore.

This way I'm already able to populate a vector of scalars. It is still incomplete, e.g. vector of structs is not yet supported, and there is quite a lot of trial & error. But as I also already considered to ask here, I though to share my approach here. If you're interested I'll provide more details. And as @frankist said: if I miss a more obvious and cleaner way to achieve this, I'm courious to learn about it.

**Reply by frankist** [reply]
Thanks @senr80, for sharing your solution! However, as you said, your solution also seems to have the same drawbacks related to the lack of "help" documentation automatically generated by CLI11.

Maybe @phlptp knows a cleaner way. Let's see what he has to say.

**Comment by frankist** [comment]
After some further analysis of the CLI11 lib, I couldn't find a clean way to support lists of objects using just options, subcommands or option groups. I assume that, at this stage, the feature is not supported. This is likely because the lib is primarily designed with command-line argument parsing in mind (rather than config file parsing) and lists of objects are something hard to configure via the command line.

I guess the follow-up question would be if there is an easy way to customize the format output for a single option (in this case, my list of objects). My idea would be that, in the case that the option is a list of objects, I could have a custom formatter that prints each subfield of the object type.

**Comment by phlptp** [comment]
I poked at it a little,  When I have had config files that had objects what I did in the past was ran all the normal flags and fields  through CLI11.   Then just parsed the file directly to extract the array of structures either as an entirely separate parse of CLI11 or just completely outside of it.    There isn't really a good way of handing arrays of identical structures in CLI11 at present.

**Reply by frankist** [reply]
Thanks for the confirmation!
